installer/cmd/render-mirror.go

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/docker/distribution/reference"
+	"github.com/gitpod-io/gitpod/installer/pkg/common"
+	"github.com/spf13/cobra"
+	"regexp"
+	"sigs.k8s.io/yaml"
+	"sort"
+	"strings"
+)
+
+type mirrorRepo struct {
+	Original string `json:"original"`
+	Target   string `json:"target"`
+}
+
+var mirrorOpts struct {
+	ExcludeThirdParty bool
+	Output            string
+}
+
+var outputTypes = []string{"yaml", "json"}
+
+// mirrorCmd represents the mirror command
+var mirrorCmd = &cobra.Command{
+	Use:   "mirror",
+	Short: "Renders a list of images used so they can be mirrored to a third-party registry",
+	Long: `Renders a list of images used so they can be mirrored to a third-party registry
+
+A config file is required which can be generated with the init command.
+
+The output can then be used to iterate over each image. A script can
+be written to pull from the "original" path and then tag and push the
+image to the "target" repo`,
+	Example: `
+  # YAML
+  gitpod-installer render mirror --config config.yaml > mirror.yaml
+
+  # JSON
+  gitpod-installer render mirror --config config.yaml -o json > mirror.json
+
+  # Pull original and push to target
+  for row in $(gitpod-installer render mirror --config ./config.yaml ./versions.yaml -o json | jq -c '.[]'); do
+    original=$(echo $row | jq -r '.original')
+    target=$(echo $row | jq -r '.target')
+    docker pull $original
+    docker tag $original $target
+    docker push $target
+  done`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		// Validate output type
+		var output string
+		for _, i := range outputTypes {
+			if mirrorOpts.Output == i {
+				output = i
+			}
+		}
+		if output == "" {
+			return fmt.Errorf("unknown output type: %s", mirrorOpts.Output)
+		}
+
+		_, cfgVersion, cfg, err := loadConfig(renderOpts.ConfigFN)
+		if err != nil {
+			return err
+		}
+
+		// Get the target repository from the config
+		targetRepo := strings.TrimRight(cfg.Repository, "/")
+
+		// Use the default Gitpod registry to pull from
+		cfg.Repository = common.GitpodContainerRegistry
+
+		k8s, err := renderKubernetesObjects(cfgVersion, cfg)
+		if err != nil {
+			return err
+		}
+
+		// Map of images used for deduping
+		allImages := make(map[string]bool)
+
+		rawImages := make([]string, 0)
+		for _, item := range k8s {
+			rawImages = append(rawImages, getPodImages(item)...)
+			rawImages = append(rawImages, getGenericImages(item)...)
```

Comment:
Just to understand what happens: We are looking in the generated k8s output for everything that looks like an image, right?

When we install Gitpod, is there a place where we replace all these images with our mirrored ones? Or is that still to come?

Reply:
Yes, this has been done. The [common.ImageName](https://github.com/gitpod-io/gitpod/blob/93af1c2a2cc8c8dee7c290e8a74e3f508a4274bc/installer/pkg/common/common.go#L339) function is used for images where we specify the container path as part of the called - this is the case for all Gitpod images, but also a few others (eg, the `kube-rbac-proxy` image).

There is also the [ThirdPartyContainerRepo](https://github.com/gitpod-io/gitpod/blob/93af1c2a2cc8c8dee7c290e8a74e3f508a4274bc/installer/pkg/common/common.go#L517) function if it's something which may be a non-Gitpod registry or uses the registry if it's mirrored. The default value can legitimately be `""` as it will default to `docker.io`, such as [here](https://github.com/gitpod-io/gitpod/blob/93af1c2a2cc8c8dee7c290e8a74e3f508a4274bc/installer/pkg/components/ws-scheduler/configmap.go#L86)

Reply:
OK, I see when I change `repository` in the config, the third-party images are fetched from this repos as well.

However, I wonder what the use case for the `--exclude-third-party` flag would look like? This flag would only make sense when the users are able to tell the installer that all Gitpod images should be fetched from their private repo but the third-party images from the original repo, wouldn't it? Or is it more that the users have a mirror for the third-party images already in place but need to add the Gitpod images to their repo as well? Is that actually a possible use case?

Reply:
Yes, talking with some of the community, it appears that there is a desire to only mirror the Gitpod images as they already mirror public registries (quay.io, docker.io etc). You're right that it will likely be an edge case, but there is a legitimate use case for it (and repo mirroring is an edge case anyway)